File: androidApp/src/main/java/com/example/justdesserts/androidApp/ui/auth/login/LoginViewModel.kt
package com.example.justdesserts.androidApp.ui.auth.login

import androidx.lifecycle.ViewModel
import com.apollographql.apollo.api.ApolloExperimental
import com.example.justdesserts.shared.repository.AuthRepository
import com.example.justdesserts.type.UserInput

@ApolloExperimental
class LoginViewModel constructor(private val repository: AuthRepository): ViewModel() {
    suspend fun signIn(email: String, password: String): String {
        return repository.signIn(UserInput(email = email, password = password))
    }
    suspend fun signUp(email: String, password: String): String {
        return repository.signUp(UserInput(email = email, password = password))
    }
    fun getAuthToken(): String {
        return repository.getUserState()?.token ?: ""
    }
    fun deleteAuthToken() {
        repository.deleteUserState()
    }
}

File: iosApp/iosApp/Features/Auth/Login/LoginViewModel.swift
//
//  LoginViewModel.swift
//  iosApp
//
//  Created by Michael Stromer on 1/26/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import shared

class LoginViewModel: ViewModel, ObservableObject {
    
    let authRepository: AuthRepository
    
    @Published public var token: String = ""
    
    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
        token = authRepository.getUserState()?.token ?? ""
    }
    
    func signIn(email: String, password: String) {
        authRepository.signIn(userInput: UserInput(email: email, password: password)) { [weak self] (data, error) in
            guard let self = self,
                  let token = data else { return }
            self.token = token
        }
    }
    
    func signUp(email: String, password: String) {
        authRepository.signUp(userInput: UserInput(email: email, password: password)) { [weak self] (data, error) in
            guard let self = self,
                  let token = data else { return }
            self.token = token
            
        }
    }
    
    func getUserState() -> UserState? {
        return authRepository.getUserState()
    }
    
    func deleteUserState() {
        authRepository.deleteUserState()
        self.token = ""
    }
}
